### PR TITLE
New rule S6961 for C#: API Controllers should derive from ControllerBase instead of Controller

### LINKED ISSUE #9090: New rule S6961 for C#: API Controllers should derive from ControllerBase instead of Controller
RSPEC: https://github.com/SonarSource/rspec/pull/3846

- [x] Basic implementation
- [x] Cover partial classes cases
- [x] Add data test methods and if needed ITs
- [x] Check if Rspec needs update

### PR DESCRIPTION
**Pull request by mary-georgiou-sonarsource**
fix #9090

### CODE REVIEW

File: analyzers/src/SonarAnalyzer.CSharp/Rules/ApiControllersShouldDeriveFromController.cs
Comment (mary-georgiou-sonarsource) [inline analyzers/src/SonarAnalyzer.CSharp/Rules/ApiControllersShouldDeriveFromController.cs:33]: is there a better way to do this? So that we can check the full namespace of where it comes from.
I think not unless I check the semantics of the identifier nodes.

Comment (antonioaversa) [reply analyzers/src/SonarAnalyzer.CSharp/Rules/ApiControllersShouldDeriveFromController.cs:33]: Just noticed that this can be made `static`.

File: analyzers/src/SonarAnalyzer.CSharp/Rules/ApiControllersShouldDeriveFromController.cs
Comment (mary-georgiou-sonarsource) [reply analyzers/src/SonarAnalyzer.CSharp/Rules/ApiControllersShouldDeriveFromController.cs:79]: Partial classes are managed in the next PR https://github.com/SonarSource/sonar-dotnet/pull/9107.

File: analyzers/src/SonarAnalyzer.CSharp/Rules/ApiControllersShouldDeriveFromController.cs
Comment (mary-georgiou-sonarsource) [reply analyzers/src/SonarAnalyzer.CSharp/Rules/ApiControllersShouldDeriveFromController.cs]: Code has been changed.

File: analyzers/src/SonarAnalyzer.CSharp/Rules/ApiControllersShouldDeriveFromController.cs
Comment (mary-georgiou-sonarsource) [reply analyzers/src/SonarAnalyzer.CSharp/Rules/ApiControllersShouldDeriveFromController.cs:93]: As discussed offline - let's be specific and check for type syntax.

File: analyzers/tests/SonarAnalyzer.Test/TestCases/ApiControllersShouldDeriveFromController.cs
Comment (mary-georgiou-sonarsource) [reply analyzers/tests/SonarAnalyzer.Test/TestCases/ApiControllersShouldDeriveFromController.cs]: It was part of the UT parametrized tests PR - I checrry picked it here as discussed.

File: analyzers/tests/SonarAnalyzer.Test/TestCases/ApiControllersShouldDeriveFromController.cs
Comment (mary-georgiou-sonarsource) [reply analyzers/tests/SonarAnalyzer.Test/TestCases/ApiControllersShouldDeriveFromController.cs:187]: I did not add the tests because base on this line https://github.com/SonarSource/sonar-dotnet/blob/a4d269158f51d9840f9dbbe9c9730944cb9ac391/analyzers/src/SonarAnalyzer.CSharp/Rules/ApiControllersShouldDeriveFromController.cs#L62
we will never reach the baseList part and I cannot inherit from two classes to force it :/

File: analyzers/src/SonarAnalyzer.CSharp/Rules/ApiControllersShouldDeriveFromController.cs
Comment (mary-georgiou-sonarsource) [reply analyzers/src/SonarAnalyzer.CSharp/Rules/ApiControllersShouldDeriveFromController.cs]: The code now has a codesmell because of this.
IMO, I do find the new version clearer, and I'd ignore this codesmell as it won't fix with a reasoning.
WDYt?

**Comment by mary-georgiou-sonarsource** [comment]
@antonioaversa, there's a complexity code smell, but from my POV, I think the way it's now is clearer.
I'd ignore it - unless you see a way to simplify the conditions and nested ifs but keep it readable and not overengineered.

File: analyzers/src/SonarAnalyzer.CSharp/Rules/ApiControllersShouldDeriveFromController.cs
Comment (antonioaversa) [reply analyzers/src/SonarAnalyzer.CSharp/Rules/ApiControllersShouldDeriveFromController.cs]: Given that `compilationStartContext` is only used to check for ASP.NET references and to `RegisterSymbolStartAction`, I would get rid of a bit of nesting by extracting the entire lambda of `RegisterSymbolStartAction` into a dedicated `CheckType(ISymbol symbol)` method. WDYT?

File: analyzers/tests/SonarAnalyzer.Test/Rules/ApiControllersShouldDeriveFromControllerTest.cs
Comment (antonioaversa) [inline analyzers/tests/SonarAnalyzer.Test/Rules/ApiControllersShouldDeriveFromControllerTest.cs:115]: Nitpick: given that there is no empty line after `#if`, I would remove the empty line before `#endif`, to make the block look symmetric.

File: analyzers/tests/SonarAnalyzer.Test/Rules/RouteTemplateShouldNotStartWithSlashTest.cs
Comment (antonioaversa) [inline analyzers/tests/SonarAnalyzer.Test/Rules/RouteTemplateShouldNotStartWithSlashTest.cs:96]: We missed that, and as a result the `RouteTemplateShouldNotStartWithSlash_HttpAttributes` test was never executed right?
If so, it seems a very good rule idea: "Methods with DataRow should be marked as DataTestMethod". WDYT, is it worth creating an issue in the backlog?

File: analyzers/tests/SonarAnalyzer.Test/Rules/ApiControllersShouldDeriveFromControllerTest.cs
Comment (antonioaversa) [inline analyzers/tests/SonarAnalyzer.Test/Rules/ApiControllersShouldDeriveFromControllerTest.cs:64]: It's not clear to me why the parameter is named `assignment`, when it's just being returned as an expression by `Foo`. I would call it something like `member`. 
If you want to have some additional cases with assignment (such as `ViewData["foo"] = "something"`, then you can call the parameter `expression`.

File: analyzers/tests/SonarAnalyzer.Test/Rules/ApiControllersShouldDeriveFromControllerTest.cs
Comment (antonioaversa) [inline analyzers/tests/SonarAnalyzer.Test/Rules/ApiControllersShouldDeriveFromControllerTest.cs:101]: I don't think it's necessary to inject `construct` as part of the class name: it doesn't play a role in determining the outcome of the test.
Instead, I would preserve the value given to it in the `DataRow` using `DisplayName`, that would give semantics to each of the rows and would make the single tests appearing better in Test Explorer.
```cs
[DataRow("object this[int index])] => View()", DisplayName = "Indexer")]
```

**Review by antonioaversa — APPROVED** [review]
LGTM!
I have left some new comments and replied to already open discussions.
Anyway, the implementation looks good and there is no blocking remark left.
So, the PR doesn't need to go back to me and can be merged after checking the remaining discussion threads, and if the proposal to reduce the complexity below 15 fixes the issue on SonarCloud.

File: analyzers/tests/SonarAnalyzer.Test/Rules/RouteTemplateShouldNotStartWithSlashTest.cs
Comment (mary-georgiou-sonarsource) [reply analyzers/tests/SonarAnalyzer.Test/Rules/RouteTemplateShouldNotStartWithSlashTest.cs:96]: @antonioaversa I'll reverse this as I fixed it here to not pile up irrelevant functional changes https://github.com/SonarSource/sonar-dotnet/pull/9114

File: analyzers/tests/SonarAnalyzer.Test/Rules/ApiControllersShouldDeriveFromControllerTest.cs
Comment (mary-georgiou-sonarsource) [reply analyzers/tests/SonarAnalyzer.Test/Rules/ApiControllersShouldDeriveFromControllerTest.cs:64]: Bad naming! I gave it a name to make the test work and then forgot to update it.

File: analyzers/src/SonarAnalyzer.CSharp/Rules/ApiControllersShouldDeriveFromController.cs
Comment (gregory-paidis-sonarsource) [inline analyzers/src/SonarAnalyzer.CSharp/Rules/ApiControllersShouldDeriveFromController.cs:24]: Isn't the name of this analyzer the opposite of what the rule suggests?
Shouldn't it be `ApiControllersShouldDeriveFromControllerBASE`?

Comment (mary-georgiou-sonarsource) [reply analyzers/src/SonarAnalyzer.CSharp/Rules/ApiControllersShouldDeriveFromController.cs:24]: Good catch!
I named it like that and then got confused as `Base` indicates a base class.
I renamed all the files to `ApiControllersShouldNotDeriveDirectlyFromController` so that it's clearer.

Comment (mary-georgiou-sonarsource) [reply analyzers/src/SonarAnalyzer.CSharp/Rules/ApiControllersShouldDeriveFromController.cs]: I prefer to keep it as is.
It has happened a lot of times that I'm searching for a *Base in the solution to actually get a base class.
So IMO it will be confusing.
Happy to be contradicted.

Comment (gregory-paidis-sonarsource) [reply analyzers/src/SonarAnalyzer.CSharp/Rules/ApiControllersShouldDeriveFromController.cs]: Sure, makes sense. :)

File: analyzers/tests/SonarAnalyzer.Test/Rules/ApiControllersShouldDeriveFromControllerTest.cs
Comment (antonioaversa) [reply analyzers/tests/SonarAnalyzer.Test/Rules/ApiControllersShouldDeriveFromControllerTest.cs]: The empty line is still there. Non blocking.

File: analyzers/tests/SonarAnalyzer.Test/Rules/AspNet/ApiControllersShouldNotDeriveDirectlyFromControllerTest.cs
Comment (antonioaversa) [inline analyzers/tests/SonarAnalyzer.Test/Rules/AspNet/ApiControllersShouldNotDeriveDirectlyFromControllerTest.cs:95]: `DisplayName = "Contrustor"` -> `DisplayName = "Constructor"`

File: analyzers/tests/SonarAnalyzer.Test/Rules/AspNet/ApiControllersShouldNotDeriveDirectlyFromControllerTest.cs
Comment (antonioaversa) [inline analyzers/tests/SonarAnalyzer.Test/Rules/AspNet/ApiControllersShouldNotDeriveDirectlyFromControllerTest.cs:96]: I am confused by the fact that the name of the destructor doesn't match the name of the class in the snippet. I think this should be `~Test() => foo = View()`. Does it compile?

**Review by antonioaversa — APPROVED** [review]
I have left a few more comments on the new code.
Once addressed/answered the PR can be merged.

File: analyzers/tests/SonarAnalyzer.Test/Rules/AspNet/ApiControllersShouldNotDeriveDirectlyFromControllerTest.cs
Comment (mary-georgiou-sonarsource) [reply analyzers/tests/SonarAnalyzer.Test/Rules/AspNet/ApiControllersShouldNotDeriveDirectlyFromControllerTest.cs]: No - but why the test is passing is the question....

Comment (antonioaversa) [reply analyzers/tests/SonarAnalyzer.Test/Rules/AspNet/ApiControllersShouldNotDeriveDirectlyFromControllerTest.cs]: Resolved offline. The `VerifyNoIssue` is not conceived to be used this way. It should be rather used when non-compliant code is under test, but the non-compliances are not raised for some reason (e.g. in a Test context, if the rule is Main).
The correct way of doing it is to use a normal `Verify` with `// Compliant` code (compliant comments are, as usual, not relevant for the verifier, and only defined for better understanding).

**Review by antonioaversa — APPROVED** [review]
LGTM!

**Pull request merged**
